Clean up REPL/LineEdit multi-line paste code

ScottPJones opened the pull request:
Use standard physical tabs for default, allow setting (in `~/.juliarc.jl` for example):
`Base.LineEdit.set_physical_tab_size(sz)`

ScottPJones committed:
Clean up REPL/LineEdit multi-line paste code

Use standard physical tabs for default, allow setting

tkelman commented:
If this is going to be configurable via a global, then yes it's much better for it to be inside `LineEdit` than to be at the top level in `Base`. But there's not much point in having an undocumented unexported setter method just to change the value of an undocumented unexported module-global integer. `Base.LineEdit.tabwidth = 8` seems good enough for this. And it should default to 4, since this is for the Julia REPL and that's the standard formatting convention for Julia code.

tkelman commented on base/LineEdit.jl:
```diff
@@ -1326,6 +1326,28 @@ function commit_line(s)
+end
+
+function bracketed_paste(s)
+    global physical_tab_size
```
doesn't need to be specifically declared like this

ScottPJones committed:
Change name to tabwidth instead of physical_tab_size, add get_tabwidth()

ScottPJones commented:
>Base.LineEdit.tabwidth = 8 seems good enough for this

Unfortunately, for some reason I don't understand, that doesn't work, which is why I added the setter.
```julia
julia> Base.LineEdit.tabwidth
8

julia> Base.LineEdit.tabwidth = 4
ERROR: type Module has no field tabwidth
```
>But there's not much point in having an undocumented unexported setter method just to change the value of an undocumented unexported module-global integer

I'd tried to add the documentation, but it wasn't being picked up (maybe that has to wait for @one-more-minute 's nice changes to get merged in).  I put it back in, but it still doesn't show up with `?`. :disappointed: 

>And it should default to 4, since this is for the Julia REPL and that's the standard formatting convention for Julia code

No, 4 is the default *indentation* in Julia, *not* the physical tab size (not even in Julia).
See https://github.com/JuliaLang/julia/commit/6c71b13e607a4bce8b02154982dbb13dc7f29dce#commitcomment-12426114
@JeffBezanson also reverted the change setting the default to 4, not 8. (see https://github.com/JuliaLang/julia/commit/85bbc1f634f19bd244831f545a1c3f8f20a0acd9)

I have also added a getter function, so that the *global* will remain undocumented/unexported, while I think we definitely would document the ability to set the REPL/LineEdit width of tabs in *paste*, for people who have editors that play tricks with tab settings (instead of having smart tabs).
Try doing `more foo.jl` on a file that is formatted with tabs set with a width of 4, it is *not* pretty.
`print` in Julia also uses 8 spaces (always?  I didn't see any way of changing that).

```julia
Scotts-Retina-MBP:julia scott$ usr/bin/julia
               _
   _       _ _(_)_     |  A fresh approach to technical computing
  (_)     | (_) (_)    |  Documentation: http://docs.julialang.org
   _ _   _| |_  __ _   |  Type "help()" for help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 0.4.0-dev+6451 (2015-08-01 02:01 UTC)
 _/ |\__'_|_|_|\__'_|  |  spj/editdef/d8bcd73* (fork: 1 commits, 0 days)
|__/                   |  x86_64-apple-darwin15.0.0

julia> print("1234567890\n\t^")
1234567890
	^
julia> Base.LineEdit.get_tabwidth()
8

julia> macro unexportedwarn(mname, replace, dispargs)
          oldname = Expr(:quote, mname)
          newname = Expr(:quote, replace)
          meta = Expr(:meta, :noinline)
          Expr(:toplevel,
              :(function $mname(args...)
                  $meta
                  Base.depwarn(string("Base.", $oldname, " was undocumented and unexported,\n",
                               "and has been removed.  Use ", $newname, "(", $dispargs, ") instead,\n",
                               "however it is also undocumented and unexported, and may change in the future."),
                               oldname)
                  $replace(args...)
              end))
       end

julia> Base.LineEdit.tabwidth = 4
ERROR: type Module has no field tabwidth

julia> Base.LineEdit.set_tabwidth!(4)
8

julia> macro unexportedwarn(mname, replace, dispargs)
          oldname = Expr(:quote, mname)
          newname = Expr(:quote, replace)
          meta = Expr(:meta, :noinline)
          Expr(:toplevel,
          :(function $mname(args...)
              $meta
              Base.depwarn(string("Base.", $oldname, " was undocumented and unexported,\n",
                   "and has been removed.  Use ", $newname, "(", $dispargs, ") instead,\n",
                   "however it is also undocumented and unexported, and may change in the future."),
                   oldname)
              $replace(args...)
          end))
       end
```

ScottPJones commented:
Hmm, GitHub's Markdown somehow messed that up... is there no way of having a "raw" quote?
Simply do `print("1234567890\n\t^")` and the ^ ends up under the 9, not the 5.

StefanKarpinski commented:
This seems good although the getter/setter thing seems over engineered, but that's ok. It seems to me that people editing Julia code are most likely to have their editor set so that tabs have width 4, which means this is most likely to work correctly with `tabwidth = 4`.

ScottPJones commented:
@StefanKarpinski I don't know why I got the error trying to just set the value:
```julia
Julia> Base.LineEdit.tabwidth = 4
ERROR: type Module has no field tabwidth
```
That's why I initially added a setter (and then decided to add a getter, so that if somebody tells me a better way to store that value, other than in a global, I can change it, and nobody would be the wiser).

I know @JeffBezanson isn't setting their physical tab width to 4, and probably nobody using Emacs either.  (you set the indentation to 4, which you can also do on VIM, and use smart tabs).
People working on the core of julia will probably also have the setting to expand tabs to spaces set anyway.
Also, it's simply non-standard.  Look at the doc's for the commands "more" and "tabs" for example on a Unix system, CSS3 standard, etc.  As I showed above, when you print with Julia, you get tabs every 8 positions (and I don't even know how you'd change that).

ScottPJones commented:
(and changing it to 4 would represent a breaking change for anybody currently using the multi-line paste,
I've at least now made it possible for people who do play tricks with physical tab size to set one single line in their juliarc to set it to 4 and be happy [and they'd probably have to also configure whatever terminal used with Julia, and the tabs command if on Mac/Unix/Linux, etc.])

ScottPJones commented:
Bump?

StefanKarpinski commented:
Merged. Note that I removed the getter and setter methods and instead documented how to modify `tabwidth` directly from outside of the module using `eval`.

ScottPJones commented:
Thanks!  Removing the getter/setter is fine by me, I just didn't know that `eval` trick.
